mu/virtual_environment.py

```diff
@@ -464,6 +474,34 @@ def quarantine_failed_venv(self):
         else:
             logger.info("Quarantined %s as %s", self.path, error_dirpath)
 
+    def recreate(self):
+        """Recreate this virtual environment with updated baseline packages and the
+        same user packages
+
+        The intended use is when the Mu version changes as this can bring with it
+        additional and/or updated packages. The simplest thing to do is to switch
+        to a new venv and then pull in the packages the user had additionally installed
+        """
+        _, user_packages = self.installed_packages()
+        #
+        # Create a new virtual environment which will then become
+        # the current one. Install the (possibly updated) baseline packages
+        #
+        self.relocate(self._generate_dirpath())
+        self.create()
+
+        #
+        # Now reinstall the original user packages
+        #
+        logger.debug("About to reinstall user packages: %s", user_packages)
+        self.install_user_packages(user_packages)
+
+        #
+        # Finally, quarantine the previous virtual environment with
+        # a marker to say it's been superseded
+        #
+        self.quarantine_venv(reason="SUPERSEDED")

```

Comment:
Hasn't `self.relocate(self._generate_dirpath())` a few lines up changed this instance to be a new virtualenv? If that's the case, isn't this quarantining the newly created venv?

Reply:
Right, but looking a few lines up, from `self.relocate(self._generate_dirpath())`:
- The `self._generate_dirpath()` generates the string for the new venv path (let's call it `/path/to/my_newly_crete_venv`)
- `self.relocate()` does `self.path = "/path/to/my_newly_crete_venv"` to this new value
- `self.create()` creates the venv stored in `self.path`
- `self.install_user_packages(user_packages)` installs all the user packages
- And then this line quarantines `self.path`? So it will go from `/path/to/my_newly_crete_venv` to `/path/to/my_newly_crete_venv.SUPERSEDED`. How does it quarantine the previous venv?

Reply:
Damn. I think you're right. I'll quarantine first, but any failure to quarantine should simply log and move on, so even if there's some kind of virus-checking issue, we should still create the new venv.

Reply:
Changes